bob/src/cluster/quorum.rs

```diff
@@ -351,25 +352,194 @@ impl Cluster for Quorum {
     }
 
     async fn exist(&self, keys: &[BobKey]) -> Result<Vec<bool>, Error> {
-        let keys_by_nodes = group_keys_by_nodes(&self.mapper, keys);
-        debug!(
-            "EXIST Nodes for fan out: {:?}",
-            &keys_by_nodes.keys().flatten().collect::<Vec<_>>()
-        );
         let len = keys.len();
-        let mut exist = vec![false; len];
-        for (nodes, (keys, indexes)) in keys_by_nodes {
-            let cluster_results = LinkManager::exist_on_nodes(&nodes, &keys).await;
-            for result in cluster_results.into_iter().flatten() {
-                for (&r, &ind) in result.inner().iter().zip(&indexes) {
-                    exist[ind] |= r;
+        debug!("EXIST {} keys", len);
+
+        let mut result = vec![false; len];
+
+        let (local, mut primary, mut secondary) = group_by_nodes(keys, &self.mapper);
+
+        if let Some(local) = local {
+            if !local.is_empty() {
+                let local_keys = local.collect(keys);
+                let local_exist = exist_on_local_node(&self.backend, &local_keys).await?;
+                local.update_existence(&mut result, &local_exist);
+            }
+        }
+
+        collect_remote_exists(&mut result, keys, &mut primary).await?;
+        collect_remote_exists(&mut result, keys, &mut secondary).await?;
+
+        let alien_index_map = IndexMap::where_not_exists(&result);
+
+        if !alien_index_map.is_empty() {
+            let local_alien_result =
+                exist_on_local_alien(&self.backend, &alien_index_map.collect(keys)).await?;
+            alien_index_map.update_existence(&mut result, &local_alien_result);
```

Comment:
Call `alien_index_map.retain_not_existed()` to remove the values, that passed the check in local aliens. This will prevent unneeded remote aliens check later